tensorflow/compiler/tf2tensorrt/convert/ops/binary_ops.cc

```diff
@@ -0,0 +1,129 @@
+/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if GOOGLE_CUDA && GOOGLE_TENSORRT
+
+#include "tensorflow/compiler/tf2tensorrt/convert/op_converter_registry.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/ops/layer_utils.h"
+
+namespace tensorflow {
+namespace tensorrt {
+namespace convert {
+const binaryOperationMap *BinaryOperationMap() {
+  static auto *const m = new binaryOperationMap({
```

Comment:
Where we will release this memory?

Reply:
Actually, this memory is never released. I just followed the pattern previously used for [`UnaryOperation`](https://github.com/tensorflow/tensorflow/pull/55082/files#diff-e911ebd6321178662cc770f8a5bd2e7262b3a8d1662004fe68b7477c7e0e237bL3731) (see lines 3731-3636)
```
const std::unordered_map<string, nvinfer1::UnaryOperation>*
UnaryOperationMap() {
  static auto* const m =
      new std::unordered_map<string, nvinfer1::UnaryOperation>({
        {"Neg", nvinfer1::UnaryOperation::kNEG},
            {"Exp", nvinfer1::UnaryOperation::kEXP},
```
and for [ActivationType](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L3204-L3217).

Is it OK?

Reply:
It was my fault to let it slip in. Would you please submit another PR to fix the existing cases?

You may mimic [here](https://github.com/tensorflow/tensorflow/blob/bc5315da4f34bd9ec0a52654978cb15da9d522a7/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L156-L164)
(1) create a routine that "new" a map
(2) use call_once to call that routine, and put the returned pointer inside a unique pointer

Reply:
OK, I will work on that PR. But I think that it would be better to create it after that one and PRs [#55229](https://github.com/tensorflow/tensorflow/pull/55229/), [#55319](https://github.com/tensorflow/tensorflow/pull/55319) will be merged.

Reply:
@bixia1: It looks like we don't need to change anything in 
```
const std::unordered_map<string, nvinfer1::UnaryOperation>*
UnaryOperationMap() {
```
or similar functions for the other maps. At least the implementation of your suggestion **create a routine that "new" a map** won't change anything from what we have now.

I just ran the following experiment. I added the output of the **m** pointer into **/Test/ccc.txt** file :
```
const std::unordered_map<string, nvinfer1::UnaryOperation>*
UnaryOperationMap() {
  static auto* const m =
      new std::unordered_map<string, nvinfer1::UnaryOperation>({
        {"Neg", nvinfer1::UnaryOperation::kNEG},
            {"Exp", nvinfer1::UnaryOperation::kEXP},
. . .
     });
  FILE *f = fopen("/Test/ccc.txt", "a");
  fprintf(f, "I am in UnaryOperationMap m = %p\n", m);
  fclose(f);
  return m;
}
```
and launched the `ConvertUnary` test.  After that I opened `/Test/ccc.txt` and I saw that all pointers, returned by `UnaryOperationMap()` are the same:
```
root@319e3ba87abd:/opt/tensorflow# grep "m = 0x55888e076e70" /Test/ccc.txt -c
290
root@319e3ba87abd:/opt/tensorflow# grep "m = " /Test/ccc.txt -c
290
root@319e3ba87abd:/opt/tensorflow# 
```
And it's is exactly what we would have with  **a routine that "new" a map**.

I will try to find another solution without **new** involved, but I think that we can live with that one.

Reply:
@bixia1 : The easiest way to avoid "new" is the following one. By using templates:
```
template <typename T>
using operationMap = std::unordered_map<std::string, T>;
// Map of all supported UnaryOperations
typedef operationMap<nvinfer1::UnaryOperation> unaryOperationMap;
const unaryOperationMap* UnaryOperationMap();
```
we define the maps
```
unaryOperationMap kUnaryOperationMap = {{
 {"Neg", nvinfer1::UnaryOperation::kNEG},
 {"Exp", nvinfer1::UnaryOperation::kEXP},
 {"Log", nvinfer1::UnaryOperation::kLOG},

#if IS_TRT_VERSION_GE(8, 2, 0, 0)
 {"Round", nvinfer1::UnaryOperation::kROUND},
 {"Sign", nvinfer1::UnaryOperation::kSIGN}
#endif
}};
```
and use their addresses:
```
const unaryOperationMap* UnaryOperationMap() {
 return &kUnaryOperationMap;
}
```
It will be a very easy PR.

Reply:
> @bixia1 : The easiest way to avoid "new" is the following one. By using templates:
> 
> ```
> template <typename T>
> using operationMap = std::unordered_map<std::string, T>;
> // Map of all supported UnaryOperations
> typedef operationMap<nvinfer1::UnaryOperation> unaryOperationMap;
> const unaryOperationMap* UnaryOperationMap();
> ```
> 
> we define the maps
> 
> ```
> unaryOperationMap kUnaryOperationMap = {{
>  {"Neg", nvinfer1::UnaryOperation::kNEG},
>  {"Exp", nvinfer1::UnaryOperation::kEXP},
>  {"Log", nvinfer1::UnaryOperation::kLOG},
> 
> #if IS_TRT_VERSION_GE(8, 2, 0, 0)
>  {"Round", nvinfer1::UnaryOperation::kROUND},
>  {"Sign", nvinfer1::UnaryOperation::kSIGN}
> #endif
> }};
> ```
> 
> and use their addresses:
> 
> ```
> const unaryOperationMap* UnaryOperationMap() {
>  return &kUnaryOperationMap;
> }
> ```
> 
> It will be a very easy PR.

we can't use a global variable or file scope variable (static) for this, see style guide https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables.

Reply:
I am fine with fixing this in another PR.